src/backend/renderer/mod.rs

```diff
@@ -268,8 +267,61 @@ pub trait Renderer {
         F: FnOnce(&mut Self, &mut Self::Frame) -> R;
 }
 
+#[cfg(feature = "wayland_frontend")]
+#[non_exhaustive]
+/// Buffer type of a given wl_buffer, if managed by smithay
+pub enum BufferType {
+    /// Buffer is managed by the [`crate::wayland::shm`] global
+    Shm,
+    #[cfg(all(feature = "backend_egl", feature = "use_system_lib"))]
+    /// Buffer is managed by a currently initialized [`crate::backend::egl::display::EGLBufferReader`]
+    Egl,
+    /// Buffer is managed by the [`crate::wayland::dmabuf`] global
+    Dma,
+}
+
+/// Returns the *type* of a wl_buffer
+/// 
+/// Returns `None` if the type is not known to smithay
+/// or otherwise not supported (e.g. not initialized using one of smithays [`crate::wayland`]-handlers).
+#[cfg(all(feature = "wayland_frontend", feature = "backend_egl", feature = "use_system_lib"))]
+pub fn buffer_type(
+    buffer: &wl_buffer::WlBuffer,
+    egl_buffer_reader: Option<&EGLBufferReader>,
+) -> Option<BufferType> {
+    if egl_buffer_reader
+        .as_ref()
+        .and_then(|x| x.egl_buffer_dimensions(&buffer))
+        .is_some()
+    {
+        Some(BufferType::Egl)
+    } else if crate::wayland::shm::with_buffer_contents(&buffer, |_, _| ()).is_ok()
+    {
+        Some(BufferType::Shm)
+    } else {
+        None
+    }
+}
+
+/// Returns the *type* of a wl_buffer
+/// 
+/// Returns `None` if the type is not recognized by smithay or otherwise not supported.
+#[cfg(all(feature = "wayland_frontend", not(feature = "backend_egl"), not(feature = "use_system_lib")))]
```

Comment:
The way these cfg block are defined, if we have the `use_system_lib` feature enabled but not the `backend_egl`, this function does not exist.